### PR TITLE
fix: properly process If-Range headers in client requests

### PR DESCRIPTION
**Pull request by midchildan**
# Problem
ATS doesn't respond correctly to conditional range requests. According to [the spec][1], ATS must ignore the `Range` header and send back the full content when the condition in the `If-Range` header isn't satisified. ATS responds with `416 Requested Range Not Satisfiable` instead.

# How to reproduce
1. Prepare a cacheable HTTP endpoint for ATS (e.g., http://localhost/)
2. Launch ATS
3. Send a normal request for the entire content so that it'd be cached in ATS (e.g., `curl -vv http://localhost`)
4. Send a conditional range request for the cached content. Specify a weak etag so that it'd fail to meet the requirements needed to receive a partial response. (e.g., `curl -vv -H 'If-Range: W/"this_wont_match"' http://localhost`)

# Cause
ATS validates `If-Range` headers in `match_response_to_request_conditionals()`. If the condition specified in `If-Range` isn't met, `match_response_to_request_conditionals()` returns a response code of `HTTP_STATUS_RANGE_NOT_SATISFIABLE`. However, ATS can't determine whether a `HTTP_STATUS_RANGE_NOT_SATISFIABLE` response code came from `match_response_to_request_conditionals()` or the response of the original content. This confusion led ATS to return incorrect responses to conditional range requests.

# Changes made
- Move the `If-Range` header validation code out of `match_response_to_request_conditionals()` so that it would run right before processing the `Range` header. This makes it easier to determine the correct response.
- Add autest test cases to check how ATS responds to range requests.

[1]: https://datatracker.ietf.org/doc/html/rfc7233#section-3.2

### CODE REVIEW

**Comment by randall** [comment]
@ezelkow1 can you kick off ci for this or?

**Comment by ezelkow1** [comment]
[approve ci]

File: proxy/http/HttpTransactCache.cc [hunk 1]
```diff
@@ -1357,47 +1359,49 @@ HttpTransactCache::match_response_to_request_conditionals(HTTPHdr *request, HTTP
...
-      }
+  @return Whether the condition specified by If-range is met, if there is any.
+    If there's no If-range header, then true.
 
-      const char *raw_etags = response->value_get(MIME_FIELD_ETAG, MIME_LEN_ETAG, &raw_len);
+*/
+bool
+HttpTransactCache::validate_ifrange_header_if_any(HTTPHdr *request, HTTPHdr *response)
+{
+  if (!(request->presence(MIME_PRESENCE_RANGE) && request->presence(MIME_PRESENCE_IF_RANGE))) {
```
Comment (SolidWallOfCode) [inline proxy/http/HttpTransactCache.cc:1375]: Why is `MIME_PRESENCE_RANGE` checked here, when it's already been checked at both call sites?

**Comment by SolidWallOfCode** [comment]
Looks good overall. Excellent description!

File: proxy/http/HttpTransactCache.cc (hunk 1 above)
Comment (midchildan) [reply proxy/http/HttpTransactCache.cc:1375]: It's only meant as a precautionary measure. Should I turn it into an assertion, or would it be better to remove it completely?

**Comment by midchildan** [comment]
#8657 might've affected the test result, assuming that the CI does a merge with the current master branch before running the them. I'll try rebasing this PR.

**Comment by midchildan** [comment]
Sorry for the delay. I've rebased and tweaked the changes against the master branch.

It turns out #8657 did change a few things, but the same issue still persists. Now instead of `HTTP_STATUS_RANGE_NOT_SATISFIABLE`, it now responds with `HTTP_STATUS_PARTIAL_CONTENT` when it should be giving the whole response. And while the pre-#8657 code gave an incorrect response regardless of the content of the `If-Range` header, the current code only responds incorrectly when given an non-matching `If-Range` condition.

**Review by traeak — APPROVED** [review]
everything looks okay

File: proxy/http/HttpTransactCache.cc (hunk 1 above)
Comment (traeak) [reply proxy/http/HttpTransactCache.cc:1375]: you can remove with a comment if you like.

Comment (midchildan) [reply proxy/http/HttpTransactCache.cc]: I've fixed this in e0d44d7.

**Review by traeak — APPROVED** [review]
Looks good.

**Pull request merged**